drivers/usb/uhc/uhc_mcux_ehci.c

```diff
@@ -0,0 +1,427 @@
+{
+	k_heap_free(&mcux_transfer_alloc_pool, p);
+}
+#endif
+
+static void uhc_mcux_thread(void *p1, void *p2, void *p3)
+{
+	const struct device *dev = p1;
+	struct uhc_mcux_data *priv = uhc_get_private(dev);
+
+	while (true) {
+		USB_HostEhciTaskFunction((void *)(&priv->mcux_host));
+	}
+}
+
+static void uhc_mcux_isr(const struct device *dev)
+{
+	struct uhc_mcux_data *priv = uhc_get_private(dev);
+
+	USB_HostEhciIsrFunction((void *)(&priv->mcux_host));
+}
+
+/* MCUX controller dirver uses this callback to notify upper layer suspend related event */
+static usb_status_t mcux_host_callback(usb_device_handle deviceHandle,
+				       usb_host_configuration_handle configurationHandle,
+				       uint32_t eventCode)
+{
+	/* TODO: nothing need to do here currently.
+	 * It will be used when implementing suspend/resume in future.
+	 */
+	return kStatus_USB_Success;
+}
+
+static int uhc_mcux_init(const struct device *dev)
+{
+	const struct uhc_mcux_config *config = dev->config;
+	struct uhc_mcux_data *priv = uhc_get_private(dev);
+	k_thread_entry_t thread_entry = NULL;
+	usb_status_t status;
+
+	if ((priv->controller_id >= kUSB_ControllerEhci0) &&
+	    (priv->controller_id <= kUSB_ControllerEhci1)) {
+		thread_entry = uhc_mcux_thread;
+	}
+
+	if (thread_entry == NULL) {
+		return -ENOMEM;
+	}
+
+#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED
+	if (config->phy_config != NULL) {
+		USB_EhciPhyInit(priv->controller_id, 0u, config->phy_config);
+	}
+#endif
+
+	priv->dev = dev;
+	/* Init MCUX USB host driver. */
+	priv->mcux_host.deviceCallback = mcux_host_callback;
+	status = priv->mcux_if->controllerCreate(priv->controller_id, &priv->mcux_host,
+						 &(priv->mcux_host.controllerHandle));
+	if (status != kStatus_USB_Success) {
+		return -ENOMEM;
+	}
+
+	/* Create MCUX USB host driver task */
+	k_thread_create(&priv->drv_stack_data, config->drv_stack, CONFIG_UHC_NXP_THREAD_STACK_SIZE,
+			thread_entry, (void *)dev, NULL, NULL, K_PRIO_COOP(2), 0, K_NO_WAIT);
+	k_thread_name_set(&priv->drv_stack_data, "uhc_mcux_ehci");
+
+	return 0;
+}
+
+static int uhc_mcux_enable(const struct device *dev)
+{
+	const struct uhc_mcux_config *config = dev->config;
+
+	config->irq_enable_func(dev);
+
+	return 0;
+}
+
+static int uhc_mcux_disable(const struct device *dev)
+{
+	const struct uhc_mcux_config *config = dev->config;
+
+	config->irq_disable_func(dev);
+
+	return 0;
+}
+
+static int uhc_mcux_shutdown(const struct device *dev)
+{
+	const struct uhc_mcux_config *config = dev->config;
+	struct uhc_mcux_data *priv = uhc_get_private(dev);
+
+	config->irq_disable_func(dev);
+	k_thread_abort(&priv->drv_stack_data);
+	priv->mcux_if->controllerDestory(priv->mcux_host.controllerHandle);
+
+	return 0;
+}
+
+/* Signal bus reset, 50ms SE0 signal */
+static int uhc_mcux_bus_reset(const struct device *dev)
+{
+	int ret;
+
+	ret = uhc_mcux_bus_control(dev, kUSB_HostBusReset);
+	if (ret == 0) {
+		ret = uhc_mcux_bus_control(dev, kUSB_HostBusRestart);
```

Comment:
This should not be necessary, it will end up in an endless loop of resets because it generates a new UHC_EVT_DEV_CONNECTED_* event. (Testing your PR, I also found a bug in usbh_shell.c, will fix it ASAP.)

Reply:
Updated.